test/python/synthesis/test_siswap_synthesis.py

```diff
@@ -0,0 +1,137 @@
+    def test_siswap_corners_3_use(self, corner):
+        """Test synthesis of some 3-use special corner cases."""
+        u = Operator(corner)
+        decomposer = SiSwapDecomposer(euler_basis=["rz", "ry"])
+        circuit = decomposer(u)
+        self.assertEqual(circuit.count_ops().get("siswap", 0), 3)
+        self.assertEqual(Operator(circuit), Operator(u))
+
+    @combine(
+        corner=[
+            iSwapGate(),
+            CXGate(),
+            CXGate().power(-1 / 2),
+        ]
+    )
+    def test_siswap_corners_2_uses(self, corner):
+        """Test synthesis of some 2-use special corner cases."""
+        u = Operator(corner)
+        decomposer = SiSwapDecomposer(euler_basis=["u"])
+        circuit = decomposer(u)
+        self.assertEqual(circuit.count_ops().get("siswap", 0), 2)
+        self.assertEqual(Operator(circuit), Operator(u))
+
+    @combine(
+        corner=[
+            iSwapGate().power(1 / 2),
+            iSwapGate().power(-1 / 2),
+        ]
+    )
+    def test_siswap_corners_1_use(self, corner):
+        """Test synthesis of some 1-use special corner cases."""
+        u = Operator(corner)
+        decomposer = SiSwapDecomposer(euler_basis=["u"])
+        circuit = decomposer(u)
+        self.assertEqual(circuit.count_ops().get("siswap", 0), 1)
+        self.assertEqual(Operator(circuit), Operator(u))
+
+    @combine(corner=[Operator(IGate()) ^ Operator(IGate()), Operator(IGate()) ^ Operator(XGate())])
+    def test_siswap_corners_0_use(self, corner):
+        """Test synthesis of some 0-use special corner cases."""
+        u = Operator(corner)
+        decomposer = SiSwapDecomposer(euler_basis=["u"])
+        circuit = decomposer(u)
+        self.assertEqual(circuit.count_ops().get("siswap", 0), 0)
+        self.assertEqual(Operator(circuit), Operator(u))
+
+    @combine(
+        p=[
+            [np.pi / 4, np.pi / 6, -np.pi / 12],  # red and blue intersection (bottom face)
+            [np.pi / 4, np.pi / 6, np.pi / 12],  # inverse of above (and locally equivalent)
+            [np.pi / 4, np.pi / 8, np.pi / 8],  # half-way between CX and SWAP (peak red polytope)
+            [np.pi / 4, np.pi / 8, -np.pi / 8],  # inverse of above (and locally equivalent)
+            [np.pi / 4, np.pi / 16, np.pi / 16],  # quarter-way between CX and SWAP
+            [np.pi / 4, np.pi / 16, np.pi / 16],  # inverse of above (and locally equivalent)
+            [np.pi / 6, np.pi / 8, np.pi / 24],  # red and blue and green intersection
+            [np.pi / 6, np.pi / 8, -np.pi / 24],  # inverse of above (not locally equivalent)
+            [np.pi / 16, np.pi / 24, np.pi / 48],  # red and blue and purple intersection
+            [np.pi / 16, np.pi / 24, -np.pi / 48],  # inverse of above (not locally equivalent)
+            [np.pi / 6, np.pi / 12, -np.pi / 16],  # inside red polytope
+        ]
+    )
+    def test_siswap_special_points(self, p):
```

Comment:
It's really good to see this test - also great to have very explicit testing of a load of points we know to be special.